### PR TITLE
[Draft] Enable SVDF op

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This draft enables SVDF op. 

Related issue #8203

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

**Comment by BalyshevArtem** [comment]
To simplify review, the draft is divided into commits.
Also I will add more tests in next commits.

File: compiler/luci/lang/include/luci/IR/Nodes/CircleDanglingNode.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * @brief Virtual CircleDanglingNode in Circle used as dangling input to CircleNode
+ */
+class CircleDanglingNode final
+  : public FixedArityNode<0, CircleNodeImpl<CircleOpcode::CIRCLEDANGLINGNODE>>
+{
+public:
+  CircleDanglingNode() = default;
+};
```
Comment (BalyshevArtem) [inline compiler/luci/lang/include/luci/IR/Nodes/CircleDanglingNode.h:31]: This new virtual node helps with the import of SVDF op. 
SVDF has one specific input tensor, that doesn't store any data (its buffer is empty), but is used in the kernel (as a scratch tensor) and stores quantization parameters, type and shape.
That is why we use this new dangling node to store tensor attribute, and use it as input node to SVDF.

**Review by BalyshevArtem — COMMENTED** [review]
For reviewers

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  const std::vector<int32_t> &inputs = op.inputs;
+  const std::vector<int32_t> &outputs = op.outputs;
+  const auto tensors = context->reader()->tensors();
+  const auto opcodes = context->reader()->opcodes();
+  int32_t scratch_index;
+
+  /*
+   * доделал драфт svdf операции, щас тут какая-то беда с nnas formatom
+   * у меня уже такое было
+   */
```
Comment (m-bronnikov) [inline compiler/luci/import/src/Nodes/CircleSVDF.cpp:58]: Don't forget to remove this lines =)

Comment (BalyshevArtem) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp]: Oh, sorry)

**Comment by m-bronnikov** [comment]
Something goes wrong with build after rebase =(

File: compiler/luci/lang/include/luci/IR/CircleNodes.lst [hunk 3]
```diff
@@ -134,6 +135,7 @@ CIRCLE_NODE(INSTANCE_NORM, CircleInstanceNorm)
 // Virtual node(s)
 CIRCLE_VNODE(CIRCLEBIDIRECTIONAL_SEQUENCE_LSTM_OUT, CircleBidirectionalSequenceLSTMOut)
 CIRCLE_VNODE(CIRCLECONST, CircleConst)
+CIRCLE_NODE(CIRCLEDANGLINGNODE, CircleDanglingNode)
```
Comment (m-bronnikov) [inline compiler/luci/lang/include/luci/IR/CircleNodes.lst:138]:
```suggestion
CIRCLE_VNODE(CIRCLEDANGLINGNODE, CircleDanglingNode)
```

Comment (m-bronnikov) [reply compiler/luci/lang/include/luci/IR/CircleNodes.lst:138]: Probably, this is the reason of build fail.

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ */
+
+CircleNode *CircleSVDFBuilder::build(const circle::OperatorT &op,
+                                     GraphBuilderContext *context) const
+{
+  assert(context != nullptr);
+
+  auto graph = context->graph();
+  const std::vector<int32_t> &inputs = op.inputs;
+  const std::vector<int32_t> &outputs = op.outputs;
```
Comment (m-bronnikov) [inline compiler/luci/import/src/Nodes/CircleSVDF.cpp:50]: (optional) Could you please use `auto` when possible? I suggest it because `auto` may help to reduce code changes in future if we need to refactor something (e.g. use `circle::Operator *` instead `circle::OperatorT &`).

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  if (inputs.size() != args.op.inputs.size())
+    return false;
+
+  return true;
+}
+
+/**
+ * @brief  SVDF Node builder
+ *
+ * @note
```
Comment (m-bronnikov) [inline compiler/luci/import/src/Nodes/CircleSVDF.cpp:40]: e.g. missed `note` comment

**Review by m-bronnikov — DISMISSED** [review]
i didn't look very closely, because there are a lot of code, but itself looks good to me if all test passed and result is correct. There exist some minor comments, but better to detect and fix them in separated PRs.

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  }
+  input: "ifm"
+  input: "weight_feature"
+  input: "weight_time"
+  input: "bias"
+  input: "input_activation_state"
+  output: "ofm"
+}
+input: "ifm"
+output: "ofm"
```
Comment (m-bronnikov) [inline res/TensorFlowLiteRecipes/SVDF_000/test.recipe:61]: Q. Did you check this tflrecipe produce correct `SVDF` model.

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp (hunk 4 above)
Comment (BalyshevArtem) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp:50]: Okey, I will change it in separated PRs for this draft.

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe (hunk 6 above)
Comment (BalyshevArtem) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:61]: There are some difficulties with the last input for this SVDF op. How I mentioned before SVDF has one specific input tensor, that doesn't store any data (its buffer is empty), but is used in the kernel (as a scratch tensor) and stores quantization parameters, type and shape. And that is why I am not sure that everything is fine with this part of SVDF op. 
This recipe can produce svdf model and this model can be processed be luci-interpreter. 
If someone can tell me how i can check the correctness of this svdf model, I will be very grateful.

Comment (m-bronnikov) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:61]: I built this draft locally, converted this recipe to tflite using:

```
$ ./build/compiler/tflchef/tools/file/tflchef-file res/TensorFlowLiteRecipes/SVDF_000/test.recipe model.tflite
```

Then, I tried to infer this model using following script:

<details>
<summary> invoke_tflite.py </summary>

```python
import tensorflow as tf
import numpy as np

model_path = "./model.tflite"

with open(model_path, 'rb') as binf:
    interp = tf.lite.Interpreter(model_content=binf.read())

interp.allocate_tensors()

for idx, detail in enumerate(interp.get_input_details()):
    tensor_idx = detail['index']
    # generate random data
    tensor_data = np.random.randn(*detail['shape']).astype(detail['dtype'])
    # dump to file (if needed)
    np.save("input" + str(idx) + ".npy", tensor_data)
    # set data to interpreter
    interp.set_tensor(tensor_idx, tensor_data)

# execute model
interp.invoke()

# collect outputs to list
ouptuts = []
for idx, detail in enumerate(interp.get_output_details()):
    tensor_idx = detail['index']
    tensor_data = interp.get_tensor(tensor_idx)
    ouptuts.append(tensor_data)
    # dump to file (if needed)
    np.save("output" + str(idx) + ".npy", tensor_data)
```

</details>

But I got segfault (sorry for Russian error msg):

```
$ python3 invoke_tflite.py 
Ошибка сегментирования (стек памяти сброшен на диск)
```

My env is following:

```
$ python3
Python 3.6.9 (default, Oct  8 2020, 12:12:24) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> tf.__version__
'2.2.0'
```

@BalyshevArtem Could you please take a look? Did I miss something? May be I need to use another tflite interpreter or version of tf?

Comment (m-bronnikov) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:61]: Please, note I tried same trick but with *kws_cut* model from https://github.com/Samsung/ONE/issues/8203#issuecomment-1000912620 and with model  [cut_svdf.zip](https://github.com/Samsung/ONE/files/7794893/cut_svdf.zip) (I forgot where I found this single-op model). Execution ends successfully.

File: compiler/luci/lang/include/luci/IR/Nodes/CircleDanglingNode.h (hunk 1 above)
Comment (seanshpark) [reply compiler/luci/lang/include/luci/IR/Nodes/CircleDanglingNode.h:31]: If you introduce this virtual node, you must change all codes that handle virtual nodes.

**Comment by seanshpark** [comment]
You didn't enable for tfldump nor circledump.

File: compiler/luci/import/src/CircleReader.cpp [hunk 7]
```diff
@@ -66,10 +66,14 @@ const char *tensor_name(const circle::Tensor *tensor)
   assert(tensor != nullptr);
 
   static const char *kEmptyTensorName = "(noname)";
-  const auto tensor_name = tensor->name()->c_str();
 
-  if (!std::string(tensor_name).empty())
-    return tensor_name;
+  if (tensor->name() != nullptr)
```
Comment (seanshpark) [inline compiler/luci/import/src/CircleReader.cpp:70]: Why did you change this? name shouldn't be null.

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe (hunk 6 above)
Comment (BalyshevArtem) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:62]: @m-bronnikov, 
Thanks for the help.
As I understood the problem was in the incorrect generation of this SVDF operation, it was necessary to add the is_variable field in svdf test.recipe. I added it in the last commit. I checked your script (but I used  tensorflow 2.6.0) and no more errors are found.

File: compiler/luci/import/src/CircleReader.cpp (hunk 7 above)
Comment (BalyshevArtem) [reply compiler/luci/import/src/CircleReader.cpp]: The original model I experimented on had no names, so I added these changes and forgot to remove them from the draft. I have removed these changes.

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe (hunk 6 above)
Comment (m-bronnikov) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:62]: I checked too and no errors occured. I guess it works fine now. Thank you!

**Comment by BalyshevArtem** [comment]
> You didn't enable for tfldump nor circledump.

@seanshpark,
Could you, please, clarify what should I add in tfldump and in circledump? At this point I've added in `compiler/circledump/src/OpPrinter.cpp` and in `compiler/tfldump/src/OpPrinter.cpp` SVDFPrinter classes. 
(tfldump added in https://github.com/Samsung/ONE/pull/8231/commits/edbd69dd36a3ec9272a052e0164dcc9811a51eeb and circledump added in https://github.com/Samsung/ONE/pull/8231/commits/e0744d850683ba732999d37784a2f2f85baf1e47)

**Comment by seanshpark** [comment]
> Could you, please, clarify what should I add in tfldump and in circledump?

Current draft seems to have both changes.
Did you add these before my comment or after my comment?
I'd like to know why you want to `clarify` it. If you had the changes then I may have missed the codes. If not as you haved added these it seems OK now.

File: compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  void input_activation_state(loco::Node *node) { at(4)->node(node); }
+
+public:
+  bool asymmetric_quantize_inputs() const { return _asymmetric_quantize_inputs; }
+  void asymmetric_quantize_inputs(bool asymmetric_quantize_inputs)
+  {
+    _asymmetric_quantize_inputs = asymmetric_quantize_inputs;
+  }
+
+  int32_t svdf_rank() const { return _rank; }
```
Comment (seanshpark) [inline compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h:60]: can you please share why use `svdf_rank` not `rank` ?

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+#include <cassert>
+
+namespace luci
+{
+bool CircleSVDFBuilder::validate(const ValidateArgs &args) const
+{
+  const auto &inputs = args.op.inputs;
+
+  if (inputs.size() != args.op.inputs.size())
```
Comment (seanshpark) [inline compiler/luci/import/src/Nodes/CircleSVDF.cpp:31]: what is the meaning if this?

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    auto input_node = context->nodefinder()->node(input_tensor_index);
+    if (input_node != nullptr)
+      input_nodes.push_back(input_node);
+    else
+      scratch_index = input_tensor_index;
+  }
+
+  auto svdf_scratch_tensor = tensors.at(scratch_index);
+  auto *svdf_scratch_node = graph->nodes()->create<CircleDanglingNode>();
+  copy_tensor_attributes(svdf_scratch_tensor, svdf_scratch_node);
```
Comment (seanshpark) [inline compiler/luci/import/src/Nodes/CircleSVDF.cpp:67]: I'm not sure of this implementation. Please explain more about this.
Whatg is `svdf_scratch_tensor`?

File: compiler/luci/import/include/luci/Import/Nodes/CircleSVDF.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+#ifndef __LUCI_IMPORT_OP_CIRCLE_SVDF_H__
+#define __LUCI_IMPORT_OP_CIRCLE_SVDF_H__
+
+#include "luci/Import/GraphBuilderBase.h"
+
+namespace luci
+{
+
+class CircleSVDFBuilder : public GraphBuilderBase
```
Comment (seanshpark) [inline compiler/luci/import/include/luci/Import/Nodes/CircleSVDF.h:25]: Q) Why did you inherit from `GraphBuilderBase` ?
Is there some particular reason or you didn't look at other nodes import code?

File: compiler/luci/logex/src/FormattedGraph.cpp [hunk 12]
```diff
@@ -2030,6 +2039,21 @@ bool SummaryBuilderLet<SB::STUV>::summary(const luci::CircleSum *node, locop::No
...
+{
+  assert(node->fusedActivationFunction() != luci::FusedActFunc::UNDEFINED);
+
+  s.args().append("input", tbl()->lookup(node->input()));
+  s.args().append("weight_feature", tbl()->lookup(node->weight_feature()));
+  s.args().append("weight_time", tbl()->lookup(node->weight_time()));
+  s.args().append("bias", tbl()->lookup(node->bias()));
+  s.args().append("rank", to_str(node->svdf_rank()));
+  s.args().append("asymmetric_quantize_inputs", to_str(node->asymmetric_quantize_inputs()));
+  s.args().append("fused", to_str(node->fusedActivationFunction()));
```
Comment (seanshpark) [inline compiler/luci/logex/src/FormattedGraph.cpp:2052]: why is `input_activation_state` not shown?

File: compiler/tflchef/tflite/src/Op/SVDF.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace tflchef
+{
+
+void TFliteOpSVDF::filler(const tflite::Operator *op, TFliteImport *import,
+                          tflchef::ModelRecipe *model_recipe) const
+{
+  const std::vector<int32_t> &inputs = as_index_vector(op->inputs());
+
+  bool hasBias = (inputs.size() == 5);
```
Comment (seanshpark) [inline compiler/tflchef/tflite/src/Op/SVDF.cpp:29]: is there bias in SVDF ?

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  svdf_options {
+    rank: 1
+    activation: RELU
+    asymmetric_quantize_inputs: false
+  }
+  input: "ifm"
+  input: "weight_feature"
+  input: "weight_time"
+  input: "bias"
+  input: "input_activation_state"
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/SVDF_000/test.recipe:58]: If `input_activation_state` is optional input then please add `_001` recipe for this too.

File: compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h (hunk 8 above)
Comment (BalyshevArtem) [reply compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h:60]: `svdf_rank` naming allows not to hide the definition of non-virtual function `rank` from class `NodeMixin<NodeTrait::TensorShape>`, from which we inherit

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp (hunk 9 above)
Comment (BalyshevArtem) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp:31]: Oh, I will rewrite it, thank you :)

File: compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h (hunk 8 above)
Comment (seanshpark) [reply compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h:60]: Current draft has
```cxx
class CircleSVDF final : public FixedArityNode<5, CircleNodeImpl<CircleOpcode::SVDF>>,
                         public CircleNodeMixin<CircleNodeTrait::FusedActFunc>
```
which does not have `TensorShape`

Comment (seanshpark) [reply compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h:60]: `CircleNode` has `NodeMixin<NodeTrait::TensorShape>`

Comment (seanshpark) [reply compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h:60]: OK, we may not use `rank()` as is.
Could you please add a comment like this?
```cxx
  // NOTE rank() exist in NodeMixin<NodeTrait::TensorShape>
  int32_t svdf_rank() const { return _rank; }
```

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp (hunk 10 above)
Comment (BalyshevArtem) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp:67]: `SVDF` has one specific input tensor, that doesn't store any data (its buffer is empty), but is used in the kernel (as a scratch tensor) and stores quantization parameters, type and shape.

svdf_inputs

In order to import and save tensor parameters `GraphBuilderBase` is used. Otherwise, this specific input tensor is imported as null when using `GraphBuilder`. 

This part of code allows us not to lose the scratch tensor index.
```c++
    auto input_node = context->nodefinder()->node(input_tensor_index);
    if (input_node != nullptr)
      input_nodes.push_back(input_node);
    else
      scratch_index = input_tensor_index;
```
 `context->nodefinder()->node(input_tensor_index);` - return null for this input scratch tensor due to data buffer of this tensor is null.

After we save tensor index for this scratch tensor we can manually create a new dangling node that will simply store tensor attributes.
```c++
 auto svdf_scratch_tensor = tensors.at(scratch_index);
  auto *svdf_scratch_node = graph->nodes()->create<CircleDanglingNode>();
  copy_tensor_attributes(svdf_scratch_tensor, svdf_scratch_node);
```

File: compiler/luci/import/include/luci/Import/Nodes/CircleSVDF.h (hunk 11 above)
Comment (BalyshevArtem) [reply compiler/luci/import/include/luci/Import/Nodes/CircleSVDF.h:25]: I tried to explain here https://github.com/Samsung/ONE/pull/8231#discussion_r782008024

File: compiler/luci/logex/src/FormattedGraph.cpp (hunk 12 above)
Comment (BalyshevArtem) [reply compiler/luci/logex/src/FormattedGraph.cpp:2052]: this is typo, I will add `input_activation_state`

File: compiler/tflchef/tflite/src/Op/SVDF.cpp (hunk 13 above)
Comment (BalyshevArtem) [reply compiler/tflchef/tflite/src/Op/SVDF.cpp:29]: bias is an optional input tensor. Now in this draft the luci-interpreter works only when the SVDF operation has bias and at the import stage (luci/import) we will check that the number of input tensors is 5.

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe (hunk 14 above)
Comment (BalyshevArtem) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:58]: `input_activation_state` is not optional input.
`bias` is optional input and should I add `_001` recipe for this situation?

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp (hunk 10 above)
Comment (seanshpark) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp:67]: Thanks for the explanation. I think I need more understanding of this Op.
(1) there's `kind: Variable` which I think is new in compiler frontend.
can you please share what you know for this?
(2) For `CircleDanglingNode`: I'm not certain of introducing this Op for it's purpose and it's name; I need to understand more of this Op. If this `CircleDanglingNode` has specific purpose only for SVDF then we need to name this to reflect that purpose. Or we can consider reusing existing virtual node (like CircleOutputDummy)

Comment (seanshpark) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp:67]: Q) How can I create this op with TensorFlow or pytorch?

File: compiler/tflchef/tflite/src/Op/SVDF.cpp (hunk 13 above)
Comment (seanshpark) [reply compiler/tflchef/tflite/src/Op/SVDF.cpp:29]: If bias can be optional we have to reflect this in import module.

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe (hunk 14 above)
Comment (seanshpark) [reply res/TensorFlowLiteRecipes/SVDF_000/test.recipe:58]: > bias is optional input and should I add _001 recipe for this situation?

Yes, please :)

File: compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h (hunk 8 above)
Comment (seanshpark) [reply compiler/luci/lang/include/luci/IR/Nodes/CircleSVDF.h:60]: If you haven't applied this yet, I think this seems better
```
  // NOTE rank() is already taken by NodeMixin<NodeTrait::TensorShape>
```

File: compiler/luci/import/src/Nodes/CircleSVDF.cpp (hunk 10 above)
Comment (seanshpark) [reply compiler/luci/import/src/Nodes/CircleSVDF.cpp:67]: Q) about the word `scratch`; is this referenced or your invention? if referenced can you please share some URLs?

File: res/TensorFlowLiteRecipes/SVDF_000/test.recipe [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+  filler {
+    tag: "gaussian"
+    arg: "0.0"
+    arg: "1.0"
+  }
+}
+operand {
+  name: "input_activation_state"
+  type: FLOAT32
+  is_variable: true
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/SVDF_000/test.recipe:39]: where is `is_variable` processed?

**Comment by seanshpark** [comment]
> Current assumption in luci is that there is no variable as training is done.

@binarman , this is my comment and this can be serious for current implemtation to support this for SVDF op. How is importance of this Op? What is the model your team want to enable?

**Comment by binarman** [comment]
@seanshpark 
> What is the model (or kind of model) your team want to enable?

This is for **KeyWord Spotting** (aka KWS) model, used by Google to detect "ok google" phrase: https://github.com/Samsung/ONE/issues/8203#issue-1086756232

<details>
<summary>netron image</summary>

image
</details>

This model is used by `tflite micro` for benchmarking, so we want to support it for better comparison of `luci micro` with `tflite micro` =)

> this is my comment and this can be serious for current implemtation to support this for SVDF op

If I understand what is "variable" tensor right: It means that this tensor is a persistent between inferences.
Maybe "variable" attribute in specific luci nodes (probably `CircleDanglingNode`) will resolve this issue, but need to think some more...

**Comment by seanshpark** [comment]
> It means that this tensor is a persistent between inferences.

Thanks! I need more understanding of this...

**Comment by binarman** [comment]
> Thanks! I need more understanding of this...

At least, this is how these tensors behave =)

Also FYI:
This kind of tensors expected in LSTM and RNN kernels and there are method for resetting state to default: [link](https://www.tensorflow.org/lite/api_docs/cc/class/tflite/interpreter#resetvariabletensors)

**Comment by seanshpark** [comment]
/CC @chunseoklee @hseok-oh , does OneRT support variable tensor? (not sure this expression is correct)

**Comment by chunseoklee** [comment]
> /CC @chunseoklee @hseok-oh , does OneRT support variable tensor? (not sure 'variable tensor' expression is correct)

Short answer due to mobile github:

If input tensor is not determined in advance, it is considered 'dynamic'(maybe variable in user term). And now onert supports this feature.

**Comment by hseok-oh** [comment]
>/CC @chunseoklee @hseok-oh , does OneRT support variable tensor? (not sure 'variable tensor' expression is correct)

If `variable tensor` means that tensor used for operation's input but modified by operation, yes.  OneRT supports LSTM operation, and it uses variable tensor.

**Comment by seanshpark** [comment]
@BalyshevArtem , please post tflchef PR

**Comment by seanshpark** [comment]
This is PR sequence what I remember not to break the build (maybe some needs adjustment)
- common-artifacts add item to exclude new recipes
- res/TensorFlowLiteRecipes
- tflchef (and or circlechef if exist)
- tfldump and circledump
- luci/lang
- luci/import
- luci/service
- luci/pass
- luci/log, luci/logex
- luci/tests `addread`
- luci/exports
- luci/tests `addwrite`
- luci-interpreter
- luci/partition
- common-artifacts remove items added

**Comment by seanshpark** [comment]
Can we close this issue?
